src/memcpy_vfp.c

```diff
@@ -0,0 +1,175 @@
+                                bool buffering, bool move)
+{
+    int i, dir, block_size = BLOCK_SIZE;
+    char *tmpbuf = tmp_buf0;
+    char *src = (char *)__src;
+    char *dst = __dst;
+
+    if (0) {
+        assert(((unsigned long)src & 127) == 0);
+        assert(((unsigned long)dst & 127) == 0);
+        assert((size & 127) == 0);
+    }
+
+copy:
+    if (size < block_size)
+        block_size = size;
+
+    dir = (dst > src && move) ? -1 : 1;
+
+    if (dir < 0) {
+        src += size - block_size;
+        dst += size - block_size;
+    }
+
+    for (i = 0; i < size / block_size; i++) {
+        if (!buffering)
+            tmpbuf = dst;
+
+        tegra_aligned_block_copy_vfp(tmpbuf, src, block_size);
+
+        if (buffering)
+            tegra_aligned_block_copy_vfp(dst, tmpbuf, block_size);
+
+        src += block_size * dir;
+        dst += block_size * dir;
+    }
+
+    size -= block_size * i;
+    dst = max(dst, __dst);
+    src = max(src, __src);
+
+    if (size)
+        goto copy;
+}
+
+void tegra_memcpy_vfp(void *dst, const void *src, int size, bool src_cached)
+{
+    bool bounce_src = !!((unsigned long)src & 127);
+    bool bounce_dst = !!((unsigned long)dst & 127);
+    int left = size;
+    int block_size;
+    char *csrc = (char *)src;
+    char *cdst = dst;
+    char *__src;
+    char *__dst;
+
+    if (dst == src)
+        return;
+
+    if (!bounce_src && !bounce_dst && left > 127) {
+        tegra_do_memcpy_vfp(cdst, csrc, size & ~127, !src_cached, false);
+        csrc += size & ~127;
+        cdst += size & ~127;
+        left = size & 127;
+    }
+
+    if (left > 127) {
+        unsigned long addr, addr_aligned, offset;
+
+        if (src_cached)
+            addr = (unsigned long)cdst;
+        else
+            addr = (unsigned long)csrc;
+
+        addr_aligned = TEGRA_ALIGN(addr, 128);
+        offset       = addr_aligned - addr;
+
+        if (offset) {
+            memcpy(cdst, csrc, offset);
+
+            csrc += offset;
+            cdst += offset;
+            left -= offset;
+
+            bounce_src = !!((unsigned long)csrc & 127);
+            bounce_dst = !!((unsigned long)cdst & 127);
+        }
+    }
+
+    while (left > 127) {
+        block_size = min(BLOCK_SIZE, left & ~127);
+
+        if (bounce_src) {
+            memcpy(tmp_buf0, csrc, block_size);
+            __src = tmp_buf0;
+        } else {
+            __src = csrc;
+        }
+
+        if (bounce_dst)
+            __dst = tmp_buf0;
+        else
+            __dst = cdst;
+
+        tegra_do_memcpy_vfp(__dst, __src, block_size, false, false);
+
+        if (bounce_dst)
+            memcpy(cdst, tmp_buf0, block_size);
+
+        csrc += block_size;
+        cdst += block_size;
+        left -= block_size;
+    }
+
+    if (left)
+        memcpy(cdst, csrc, left);
+}
+
+void tegra_memmove_vfp_aligned_uncached(void *dst, void *src, int size)
+{
+    if (dst == src)
+        return;
```

Comment:
Same as above, I doubt you want to go down the path of supporting anywhere near overlapping ranges here...

Also, I think you should assert the alignment. The asserts in tegra_do_memcpy_vfp don't currently run.